### PR TITLE
plan: a pre-work for changing the cbo framework.

### PR DESCRIPTION
**Pull request by hanfei1991**
There are lots of logic can be used in new cbo framework. Two new interfaces are added:
1. `generatePhysicalPlans` to generate all physical plans, like logical join can be converted to hash join, merge join and index join.
2. `getPushedProp` tries to generate all possible properties can be pushed down.

@zimulala @winoros PTAL

### CODE REVIEW

File: plan/new_physical_plan_builder.go [hunk 1]
```diff
@@ -503,65 +462,67 @@ func (p *TopN) convert2NewPhysicalPlan(prop *requiredProp) (task, error) {
...
-		optTask, err := p.children[0].(LogicalPlan).convert2NewPhysicalPlan(&requiredProp{taskTp: taskTp})
+	// Now we only consider rootTask.
+	if len(p.basePlan.children) == 0 {
+		// When the children length is 0, we process it specially.
+		task = &rootTask{p: p.basePlan.self.(PhysicalPlan)}
+		task = prop.enforceProperty(task, p.basePlan.ctx, p.basePlan.allocator)
+		return task, p.storeTask(prop, task)
+	}
+	// Else we suppose it only has one child.
+	for _, pp := range p.basePlan.self.(LogicalPlan).generatePhysicalPlans() {
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: When the `basePlan` has multiple plans？

**Comment by hanfei1991** [comment]
@zimulala PTAL

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -53,65 +53,24 @@ func (p *requiredProp) enforceProperty(task task, ctx context.Context, allocator
...
 	newProp := &requiredProp{taskTp: rootTaskType}
-	if prop.isEmpty() {
-		return newProp, false
-	}
 	newCols := make([]*expression.Column, 0, len(prop.cols))
 	for _, col := range prop.cols {
 		idx := p.schema.ColumnIndex(col)
 		if idx == -1 {
-			return newProp, false
+			return nil
```
Comment (winoros) [inline plan/new_physical_plan_builder.go]: Should there be `nil` or `newProp`?

Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Indeed, it will never be -1

**Comment by winoros** [comment]
LGTM

**Comment by hanfei1991** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**